### PR TITLE
Support identifying interceptor url and use as API

### PR DESCRIPTION
**Pull request by dbMundada**
- Support identifying interceptor url and use as API
- Treat Graphql SDKs as APIs

### CODE REVIEW

File: src/main/scala/ai/privado/exporter/SinkExporter.scala [hunk 1]
```diff
@@ -123,6 +123,13 @@ class SinkExporter(cpg: Cpg, ruleCache: RuleCache) {
   }
 
   private def convertSinkList(sinks: List[List[Tag]]) = {
+    def isAPIUrlTag(tag: Tag): Boolean = {
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/exporter/SinkExporter.scala:126]: This looks irrelevant, this will get matched for all API sink node

File: src/main/scala/ai/privado/exporter/SinkExporter.scala [hunk 2]
```diff
@@ -171,7 +178,16 @@ class SinkExporter(cpg: Cpg, ruleCache: RuleCache) {
...
+
+    val updatedSinks = sinks.map { tags =>
+      val containsAPIUrl = tags.exists(isAPIUrlTag)
+      if (containsAPIUrl) {
+        tags.filterNot(tag => tag.value == Constants.thirdPartiesAPIRuleId)
+      } else {
+        tags
+      }
+    }
+    updatedSinks
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/exporter/SinkExporter.scala:190]: Have you tested this, doesn't look like this will work

File: src/main/scala/ai/privado/languageEngine/javascript/tagger/sink/GraphqlAPITagger.scala [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    ruleCache.getRule.sinks
+      .filter(rule => rule.nodeType.equals(NodeType.API))
+      .toArray
+  }
+
+  override def runOnPart(builder: DiffGraphBuilder, ruleInfo: RuleInfo): Unit = {
+    val commonGraphqlPackages: String = ruleCache.getSystemConfigByKey(Constants.apiGraphqlLibraries, true)
+    val graphqlREADSink: String       = ruleCache.getSystemConfigByKey(Constants.apiGraphqlReadSink, true)
+    val graphqlWRITESink: String      = ruleCache.getSystemConfigByKey(Constants.apiGraphqlWriteSink, true)
+    val cacheCall                     = cpg.call.where(_.nameNot("(<operator|<init).*")).l
```
[the remark this answers is not in the document]
Comment (khemrajrathore) [reply src/main/scala/ai/privado/languageEngine/javascript/tagger/sink/GraphqlAPITagger.scala:64]: The cacheCall variable should be outside of `runOnPart` as the reason to cache this is to run this only once

File: src/main/scala/ai/privado/languageEngine/javascript/tagger/sink/GraphqlAPITagger.scala [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    val apis                          = cacheCall.methodFullName(commonGraphqlPackages).toList
+
+    if (!ruleInfo.id.equals(Constants.internalAPIRuleId)) {
+      apis.foreach((apiNode) => {
+        val isReadAPI  = apiNode.name.matches(graphqlREADSink)
+        val isWriteAPI = apiNode.name.matches(graphqlWRITESink)
+
+        if (isReadAPI) {
+          val newRuleIdToUse = ruleInfo.id + " (Read)"
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/languageEngine/javascript/tagger/sink/GraphqlAPITagger.scala:73]: Instead of repeating ` (Read)` and ` (Write)` create a constant

File: src/main/scala/ai/privado/languageEngine/javascript/tagger/sink/JSAPITagger.scala [hunk 5]
```diff
@@ -24,16 +24,52 @@
...
 
+// class GraphqlAPITagger
 class JSAPITagger(cpg: Cpg, ruleCache: RuleCache) extends APITagger(cpg, ruleCache) {
 
   override def runOnPart(builder: DiffGraphBuilder, ruleInfo: RuleInfo): Unit = {
     super.runOnPart(builder, ruleInfo)
+
+    // baseUrl" Identify the client creation & baseUrl used
+    // TODO: Replace the "baseUrl" with already tagged sinks
+    val ClientCreationBaseUrlPattern: String =
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/languageEngine/javascript/tagger/sink/JSAPITagger.scala:59]: The variable name should be camel case or Upper case

File: src/main/scala/ai/privado/languageEngine/javascript/tagger/sink/JSAPITagger.scala [hunk 6]
```diff
@@ -24,16 +24,52 @@
...
+      ruleCache.getSystemConfigByKey(Constants.clientCreationBaseUrlPattern, true)
+    val cacheCall          = cpg.call.where(_.nameNot("(<operator|<init).*")).l
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/languageEngine/javascript/tagger/sink/JSAPITagger.scala:61]: Similar comment as above

File: src/main/scala/ai/privado/languageEngine/javascript/tagger/sink/JSAPITagger.scala [hunk 7]
```diff
@@ -24,16 +24,52 @@
...
+    val apiInternalSources = cpg.literal.code("(?:\"|'|`)(" + ruleInfo.combinedRulePattern + ")(?:\"|'|`)").l
+    val apis               = cacheCall.methodFullName(ClientCreationBaseUrlPattern).toList
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/languageEngine/javascript/tagger/sink/JSAPITagger.scala:63]: Looks to me like the `apiInternalSources`, `apis` these variables are part of the APITagger. Can we not override them?

File: src/main/scala/ai/privado/languageEngine/javascript/tagger/sink/JSAPITagger.scala [hunk 8]
```diff
@@ -24,16 +24,52 @@
...
+
+    val uniqueDomains = getBaseUrlForFrontendApps(apis, apiInternalSources)
+//    val apiCalls = cacheCall
+//      .name(APISINKS_REGEX)
+//      .methodFullNameNot(COMMON_IGNORED_SINKS_REGEX)
+//      .methodFullName(commonHttpPackages)
+//      .l
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/languageEngine/javascript/tagger/sink/JSAPITagger.scala:70]: remove commented unused code

**Comment by khemrajrathore** [comment]
@dbMundada 

- [ ] Add comments, it is hard to follow what exactly are we tagging and achieving with these changes

File: src/main/scala/ai/privado/exporter/SinkExporter.scala (hunk 2 above)
Comment (dbMundada) [reply src/main/scala/ai/privado/exporter/SinkExporter.scala]: removed will work again with different approach.

File: src/main/scala/ai/privado/languageEngine/javascript/tagger/sink/JSAPITagger.scala [hunk 9]
```diff
@@ -85,4 +133,82 @@ class JSAPITagger(cpg: Cpg, ruleCache: RuleCache) extends APITagger(cpg, ruleCac
...
+      val newRuleIdToUse = ruleInfo.id + "." + domain
+      ruleCache.setRuleInfo(ruleInfo.copy(id = newRuleIdToUse, name = ruleInfo.name + " " + domain))
+      addRuleTags(builder, apiNode, ruleInfo, ruleCache, Some(newRuleIdToUse))
+      storeForTag(builder, apiNode, ruleCache)(Constants.apiUrl + newRuleIdToUse, domain)
+    }
+
+    if (apis.nonEmpty && filteredSourceNode.nonEmpty) {
+      val apiFlows = {
+        val flows = apis.reachableByFlows(filteredSourceNode)(engineContext).toList
+        if (ScanProcessor.config.disableDeDuplication)
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/languageEngine/javascript/tagger/sink/JSAPITagger.scala:188]: Instead use `privadoInput` here

File: src/main/scala/ai/privado/languageEngine/javascript/tagger/sink/JSAPITagger.scala [hunk 10]
```diff
@@ -24,16 +24,64 @@
...
+    //      }
+    //    }
+
+    val clientCreationBaseUrlPattern: String =
+      ruleCache.getSystemConfigByKey(Constants.clientCreationBaseUrlPattern, true)
+    val apiLiterals   = cpg.literal.code("(?:\"|'|`)(" + ruleInfo.combinedRulePattern + ")(?:\"|'|`)").l
+    val initApiCalls  = cacheCall.methodFullName(clientCreationBaseUrlPattern).toList
+    val uniqueDomains = getBaseUrlForFrontendApps(initApiCalls, apiLiterals, builder, ruleInfo, ruleCache)
+    // TODO: Need another approach to map the baseUrl with actual sink nodes.
+    // sinkTaggerWithDomains(apiLiterals, initApiCalls, builder, ruleInfo, ruleCache, uniqueDomains)
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/languageEngine/javascript/tagger/sink/JSAPITagger.scala:83]: remove commented code if not required

File: src/test/scala/ai/privado/languageEngine/javascript/tagger/sink/JSAPITaggerTest.scala [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+    SystemConfig(
+      "ignoredSinks",
+      "(?i).*(?<=map|list|jsonobject|json|array|arrays|jsonnode|objectmapper|objectnode).*(put:|get:).*",
+      Language.JAVASCRIPT,
+      "",
+      Array()
+    )
+  )
+
+  val privadoInput = PrivadoInput(generateAuditReport = true)
```
Comment (khemrajrathore) [inline src/test/scala/ai/privado/languageEngine/javascript/tagger/sink/JSAPITaggerTest.scala:102]: setting generateAuditReport is not needed

**Review by khemrajrathore — APPROVED** [review]
LGTM
